src/numa_virtual.c

```diff
@@ -0,0 +1,166 @@
+            for (i=0;i<tokens->qty;i++)
+            {
+                if (binstr(tokens->entry[i],0,freeString) != BSTR_ERR)
+                {
+                     bstring tmp = bmidstr (tokens->entry[i], 10, blength(tokens->entry[i])-10  );
+                     bltrimws(tmp);
+                     struct bstrList* subtokens = bsplit(tmp,(char) ' ');
+                     free = str2int(bdata(subtokens->entry[0]));
+                     bdestroy(tmp);
+                     bstrListDestroy(subtokens);
+                }
+            }
+            bstrListDestroy(tokens);
+            bdestroy(src);
+            fclose(fp);
+        }
+    }
+    else
+    {
+        bdestroy(freeString);
+        ERROR;
+    }
+    bdestroy(freeString);
+    return free;
+}
+
+static uint64_t
+getTotalNodeMem(void)
+{
+    int i;
+    FILE *fp;
+    uint64_t total = 0;
+    bstring totalString  = bformat("MemTotal:");
+
+    if (!access("/proc/meminfo", R_OK))
+    {
+        if (NULL != (fp = fopen ("/proc/meminfo", "r")))
+        {
+            bstring src = bread ((bNread) fread, fp);
+            struct bstrList* tokens = bsplit(src,(char) '\n');
+            for (i=0;i<tokens->qty;i++)
+            {
+                if (binstr(tokens->entry[i],0,totalString) != BSTR_ERR)
+                {
+                     bstring tmp = bmidstr (tokens->entry[i], 10, blength(tokens->entry[i])-10  );
+                     bltrimws(tmp);
+                     struct bstrList* subtokens = bsplit(tmp,(char) ' ');
+                     total = str2int(bdata(subtokens->entry[0]));
+                     bdestroy(tmp);
+                     bstrListDestroy(subtokens);
+                }
+            }
+            bstrListDestroy(tokens);
+            bdestroy(src);
+            fclose(fp);
+        }
+    }
+    else
+    {
+        bdestroy(totalString);
+        ERROR;
+    }
+
+    bdestroy(totalString);
+    return total;
+}
+
+/* #####   FUNCTION DEFINITIONS  -  EXPORTED FUNCTIONS   ################## */
+
+int
+virtual_numa_init()
+{
+    int i;
+
+    numa_info.numberOfNodes = 1;
+
+    numa_info.nodes = (NumaNode*) malloc(sizeof(NumaNode));
+    if (!numa_info.nodes)
+    {
+        fprintf(stderr,"No memory to allocate %ld byte for nodes array\n",sizeof(NumaNode));
+        return -1;
+    }
+    numa_info.nodes[0].id = 0;
+    numa_info.nodes[0].numberOfProcessors = cpuid_topology.numHWThreads;
+    numa_info.nodes[0].totalMemory = getTotalNodeMem();
+    numa_info.nodes[0].freeMemory = getFreeNodeMem();
+    numa_info.nodes[0].processors = (uint32_t*) malloc(cpuid_topology.numHWThreads * sizeof(uint32_t));
+    if (!numa_info.nodes[0].processors)
+    {
+        fprintf(stderr,"No memory to allocate %ld byte for processors array of NUMA node %d\n",
+                cpuid_topology.numHWThreads * sizeof(uint32_t),0);
+        return -1;
+    }
+    for (i = 0; i < cpuid_topology.numHWThreads; i++)
+    {
+        numa_info.nodes[0].processors[i] = i;
```

Comment:
Please use `cpuid_topology.threadPool[i].apicId`, just `i` might not be right from strangely numbered hw threads.